duecredit/tests/test_io.py

```diff
@@ -112,6 +114,82 @@ def myfunction(arg42):
                 msg='value was {0}'.format(value))
     assert_equal(len(value.split('\n')), 18, msg='value was {0}'.format(value))
 
+    # test we get the reference numbering right
+    samples_bibtex = [_generate_sample_bibtex() for x in range(5)]
+    # this sucks but at the moment it's the only way to have multiple
+    # references for a function
+
+    @due.dcite(BibTeX(samples_bibtex[0]), description='another solution',
+               path='myothermodule', version='0.0.666')
+    def myothermodule(arg1, kwarg2="blah"):
+        """docstring"""
+        assert_equal(arg1, "magical")
+        assert_equal(kwarg2, 1)
+
+        @due.dcite(BibTeX(samples_bibtex[1]), description='solution to life',
+                   path='myothermodule:myotherfunction')
+        @due.dcite(BibTeX(samples_bibtex[2]), description='solution to life',
+                   path='myothermodule:myotherfunction')
+        @due.dcite(BibTeX(samples_bibtex[3]), description='solution to life',
+                   path='myothermodule:myotherfunction')
+        @due.dcite(BibTeX(samples_bibtex[4]), description='solution to life',
+                   path='myothermodule:myotherfunction')
+        # XXX: atm cross-referencing doesn't work
+        # @due.dcite(BibTeX(_sample_bibtex2), description='solution to life',
+        #           path='myothermodule:myotherfunction')
+        def myotherfunction(arg42):
+            pass
+
+        myotherfunction('argh')
+        return "load"
+
+    myothermodule('magical', kwarg2=1)
+    strio = StringIO()
+    TextOutput(strio, due).dump(tags=['*'])
+    value = strio.getvalue()
+    lines = value.split('\n')
+
+    citation_numbers = []
+    reference_numbers = []
+    references = []
+    for line in lines:
+        match_citation = re.search('(\[[0-9]\])$', line)
+        match_reference = re.search('^(\[[0-9]\])', line)
+        if match_citation:
+            citation_numbers.append(match_citation.group())
+        elif match_reference:
+            reference_numbers.append(match_reference.group())
+            references.append(line.replace(match_reference.group(), ""))
+
+    assert_equal(citation_numbers, reference_numbers)
+    assert_equal(len(set(references)), len(citation_numbers))
```

Comment:
do I get it right that all this is just to check number of citations/references? ;-)  we should make a better use of this test

and dump should get refactored into two functions -- one will prepare the "model", which we can test independently all those numbers etc; and then the other does rendering (e.g. text) of that model, where we test formatting etc